pkg/glusterfs/controllerserver.go

```diff
@@ -0,0 +1,370 @@
+
+	var glusterServer string
+	var bkpServers []string
+
+	if req == nil {
+		glog.Errorf("volume create request is nil")
+		return nil, status.Errorf(codes.InvalidArgument, "request cannot be empty")
+	}
+
+	if req.Name == "" {
+		return nil, status.Error(codes.InvalidArgument, "Name is a required field")
+	}
+	glog.V(1).Infof("creating volume with name ", req.Name)
+
+	if req.VolumeCapabilities == nil || len(req.VolumeCapabilities) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume capabilities is a required field")
+	}
+
+	// If capacity mentioned, pick that or use default size 20 MB
+	volSizeBytes := defaultVolumeSize
+	if req.GetCapacityRange() != nil {
+		volSizeBytes = int(req.GetCapacityRange().GetRequiredBytes())
+	}
+
+	volSizeMB := int(utils.RoundUpSize(volSizeBytes, 1024*1024))
+
+	// Get Volume name : TODO use the values from request
+	volumeName := req.Name
+	glusterVol := req.GetParameters()["glustervol"]
+	glusterServer = req.GetParameters()["glusterserver"]
+	glusterURL := req.GetParameters()["glusterurl"]
+	glusterURLPort := req.GetParameters()["glusterurlport"]
+	glusterUser := req.GetParameters()["glusteruser"]
+	glusterUserSecret := req.GetParameters()["glusterusersecret"]
+
+	glog.V(3).Infof("Request fields:[ %v %v %v %v %v %v]", glusterVol, glusterServer, glusterURL, glusterURLPort, glusterUser, glusterUserSecret)
+
+	// Get list of volumes in the TSP
+	volumeListResp, err := cs.client.Volumes("")
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	glog.V(2).Infof("Volume list response: %+v", volumeListResp)
+
+	/* TODO: Check the requested volume exist in the cluster
+	glusterServer, bkpServers, err = cs.checkExistingVolumes(&volumeListResp, volumeName, volSizeBytes)
+	if err == nil {
+		return &csi.CreateVolumeResponse{
+			Volume: &csi.Volume{
+				Id:            volumeName,
+				CapacityBytes: int64(volSizeBytes),
+				Attributes: map[string]string{
+					"glustervol":        volumeName,
+					"glusterserver":     glusterServer,
+					"glusterbkpservers": strings.Join(bkpServers, ":"),
+				},
+			},
+		}, nil
+	}
+	*/
+
+	// If volume does not exist, provision volume
+	glog.V(4).Infof("Received request to create/provision volume name:%s with size:%d", volumeName, volSizeMB)
+	volMetaMap := make(map[string]string)
+	volMetaMap[glusterDescAnn] = glusterDescAnnValue
+	volumeReq := api.VolCreateReq{
+		Name:     volumeName,
+		Metadata: volMetaMap,
+		Size:     uint64(volSizeMB),
+	}
+
+	glog.V(2).Infof("volume request: %+v", volumeReq)
+	volumeCreateResp, err := cs.client.VolumeCreate(volumeReq)
+	if err != nil {
+		glog.Errorf("failed to create volume : %v", err)
+		return nil, status.Error(codes.Internal, fmt.Sprintf("failed to create volume: %s", err.Error()))
+	}
+
+	glog.V(3).Infof("volume create response : %+v", volumeCreateResp)
+	err = cs.client.VolumeStart(volumeName, true)
+	if err != nil {
+		//TODO : delete the volume
```

Comment:
This todo can be handled. It's just a few lines of code and one call to delete volume.

Reply:
Would like to cover the negatives in a single commit which is tracked by https://github.com/gluster/gluster-csi-driver/issues/12

Reply:
I think we don't need to delete the volume if volume start fails
kubernetes will retry if volume creation fails as we are listing the volumes and starting it again before sending back the response